samples: matter: Introduce Timed Schedule Access in Lock

ArekBalysNordic opened the pull request:
- Regenerated Matter Lock Zap to include Timed Schedule Access feature

- Added Schedules for WeekDay, YearDay and Holiday to the credentials manager according to Matter
Specification.

- Implemented callbacks for the schedule feature required by door-lock-server.

- Moved credentials manager users, schedules and credentials implementations to separate files due to a high number of lines in the credentials_manager.cpp file.

- Added a guide to Matter Lock sample on how to use this feature.

- Updated release notes.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-15485
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

Damian-Nordic commented on samples/matter/lock/src/credentials/credentials_storage.h:
```diff
@@ -21,18 +21,38 @@
+ *        / <user index (int)> /
+ *                             / <user data (bytes)>
+ *    /sch_<type (w - weekday, y - yearday, h - holiday )>/
+ *        /schedule_usr_idxs (bytes)/
```
This doesn't matches the code.

Damian-Nordic commented on samples/matter/lock/src/credentials/credentials_manager_schedules.cpp:
```diff
@@ -0,0 +1,386 @@
+	for (size_t userIndex = 1; userIndex <= CONFIG_LOCK_MAX_NUM_USERS; userIndex++) {
+		scheduleFound = false;
+		/* Load WeekDay schedules */
+		if (!CredentialsStorage::Instance().Load(
+			    CredentialsStorage::Type::WeekDayScheduleIndexes, scheduleWeekDayIndexesSerialized,
+			    sizeof(scheduleWeekDayIndexesSerialized), outSize, userIndex)) {
+		} else {
```
```suggestion
		if (CredentialsStorage::Instance().Load(
			    CredentialsStorage::Type::WeekDayScheduleIndexes, scheduleWeekDayIndexesSerialized,
			    sizeof(scheduleWeekDayIndexesSerialized), outSize, userIndex)) {
```
or is something missing in the `if` block? Same for loading other schedules.

ArekBalysNordic replied:
I did it in the same way as it was done for users and credentials, but you are right this is not so intuitive :D I was thinking about how it works some time :D

ArekBalysNordic commented on samples/matter/lock/src/credentials/credentials_storage.cpp:
```diff
@@ -35,14 +35,17 @@ bool GetStorageFreeSpace(size_t &freeBytes)
 #define PSStore SecureStore
 #define PSLoad SecureLoad
+#else
+#error Please select only one of the possible credential backends: CONFIG_NCS_SAMPLE_MATTER_SECURE_STORAGE_BACKEND or CONFIG_NCS_SAMPLE_MATTER_SETTINGS_STORAGE_BACKEND
```
Unfortunately not, but it was done in this way that both backends can be enabled at the same time, but in this specific case, we would like to use only one.

Damian-Nordic replied:
I see, thanks.

ArekBalysNordic commented on samples/matter/lock/src/credentials/credentials_data_types.h:
```diff
@@ -118,6 +119,89 @@ struct Credential {
 
+enum class ScheduleType : uint8_t { WeekDay, YearDay, Holiday };
+
+template <ScheduleType T> struct Schedule {
```
Ok, maybe I over-thought it :D

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -481,6 +481,106 @@ Upgrading the device firmware
+   - :kconfig:option:`CONFIG_LOCK_MAX_YEARDAY_SCHEDULES_PER_USER` to define the maximum number of year-day schedules per one user.
+   - :kconfig:option:`CONFIG_LOCK_MAX_HOLIDAY_SCHEDULES` to define the maximum number of holiday schedules.
+
+To learn more how to configure Door Lock cluster see the :ref:`ug_matter_creating_accessory` user guide.
```
```suggestion
To learn more about configuring the Door Lock cluster, see the :ref:`ug_matter_creating_accessory` user guide.
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -481,6 +481,106 @@ Upgrading the device firmware
+
+To learn more how to configure Door Lock cluster see the :ref:`ug_matter_creating_accessory` user guide.
+
+All schedule timed-access entries are saved to non-volatile memory and loaded automatically after device reboot.
```
```suggestion
All scheduled timed access entries are saved to non-volatile memory and loaded automatically after device reboot.
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -481,6 +481,106 @@ Upgrading the device firmware
+To learn more how to configure Door Lock cluster see the :ref:`ug_matter_creating_accessory` user guide.
+
+All schedule timed-access entries are saved to non-volatile memory and loaded automatically after device reboot.
+To disable the feature you need to revert all changes in the :file:`lock.zap` file, re-generate the ZAP files and set the :kconfig:option:`CONFIG_LOCK_SCHEDULES` kconfig option to ``n``.
```
```suggestion
To disable the feature, you need to revert all changes in the :file:`lock.zap` file, re-generate the ZAP files and set the :kconfig:option:`CONFIG_LOCK_SCHEDULES` Kconfig option to ``n``.
```

Damian-Nordic commented on samples/matter/lock/src/credentials/credentials_manager_schedules.cpp:
```diff
@@ -71,8 +71,8 @@ DlStatus CredentialsManager<CRED_BIT_MASK>::SetWeekDaySchedule(uint8_t weekdayIn
 	if (0 < serializedSize && 0 < serializedIndexesSize) {
-		if (!CredentialsStorage::Instance().Store(CredentialsStorage::Type::WeekDaySchedule, scheduleSerialized,
-							  serializedSize, userIndex, weekdayIndex)) {
+		if (CredentialsStorage::Instance().Store(CredentialsStorage::Type::WeekDaySchedule, scheduleSerialized,
```
```suggestion
		if (!CredentialsStorage::Instance().Store(CredentialsStorage::Type::WeekDaySchedule, scheduleSerialized,
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,111 @@ After building the sample and programming it to your development kit, complete t
+
+   Both ``LocalStartTime`` and ``LocalEndTime`` are in seconds with the local time offset based on the local timezone and DST offset on the day represented by the value.
+
+   For example:
```
I'd expand this to a full sentence that explains the values used, like was done for `Week-day`.

ArekBalysNordic replied:
Done

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,111 @@ After building the sample and programming it to your development kit, complete t
+
+   .. code-block:: console
+
+      ./chip-tool doorlock set-week-day-schedule 1 1 60 120 1 1
```
```suggestion
      ./chip-tool doorlock set-year-day-schedule 1 1 60 120 1 1
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,111 @@ After building the sample and programming it to your development kit, complete t
+
+      ./chip-tool doorlock set-week-day-schedule 1 1 60 120 1 1
+
+#. Set the example ``holiday`` schedule using the following command:
```
```suggestion
#. Set the example ``Holiday`` schedule using the following command:
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,111 @@ After building the sample and programming it to your development kit, complete t
+
+#. Set the example ``holiday`` schedule using the following command:
+
+   ./chip-tool doorlock set-holiday-schedule HolidayIndex LocalStartTime LocalEndTime OperatingMode destination-id endpoint-id
```
```suggestion
   .. parsed-literal::
      :class: highlight
      
      ./chip-tool doorlock set-holiday-schedule *holiday-index* *localtime-start* *localtime-end* *operating-mode* *node-id* *endpoint-id*
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,111 @@ After building the sample and programming it to your development kit, complete t
+
+   ./chip-tool doorlock set-holiday-schedule HolidayIndex LocalStartTime LocalEndTime OperatingMode destination-id endpoint-id
+
+   Where:
+
```
```suggestion
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,111 @@ After building the sample and programming it to your development kit, complete t
+
+   Where:
+
+   * ``YearDayIndex`` is an index of the new schedule, starting from ``1``.
```
```suggestion
   * *yearday-index* is the index of the new schedule, starting from ``1``.
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,111 @@ After building the sample and programming it to your development kit, complete t
+
+   Where:
+
+   * ``HolidayIndex`` is an index of the new schedule, starting from ``1``.
```
```suggestion
   * *holiday-index* is the index of the new schedule, starting from ``1``.
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,111 @@ After building the sample and programming it to your development kit, complete t
+   * ``node-id`` is the device node ID.
+   * ``endpoint-id`` is the Matter door lock endpoint, in this sample assigned to ``1``.
+
+   For example, to setup a ``holiday-schedule`` with type ``Vacation`` use the following command:
```
Expand this to explain all the values used as well.

ArekBalysNordic replied:
Done

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,111 @@ After building the sample and programming it to your development kit, complete t
+
+      ./chip-tool doorlock set-holiday-schedule 1 60 120 1 1 1
+
+#. Read saved schedules using the following commands providing the arguments in a similar way as above:
```
```suggestion
#. Read saved schedules using the following commands and providing the same arguments you used in the earlier steps:
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -481,6 +481,111 @@ Upgrading the device firmware
+         * ``SetHolidaySchedule``
+         * ``GetHolidaySchedule``
+         * ``GetHolidayScheduleResponse``
+         * ``ClearHolidaySchedule``
+#. In the **Door Lock** context window, go to the **Attributes** tab and set the proper bits for the ``FeatureMap`` attribute:
```
```suggestion
         * ``ClearHolidaySchedule``

#. In the **Door Lock** context window, go to the **Attributes** tab and set the proper bits for the ``FeatureMap`` attribute:
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,119 @@ After building the sample and programming it to your development kit, complete t
+   * *destination-id* is the device node ID.
+   * *endpoint-id* is the Matter door lock endpoint, in this sample assigned to ``1``.
+
+   Both ``LocalStartTime`` and ``LocalEndTime`` are in seconds with the local time offset based on the local timezone and DST offset on the day represented by the value.
```
```suggestion
   Both ``localtime-start`` and ``localtime-end`` are in seconds with the local time offset based on the local timezone and DST offset on the day represented by the value.
```

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +641,119 @@ After building the sample and programming it to your development kit, complete t
+
+   * *holiday-index* is the index of the new schedule, starting from ``1``.
+   * *localtime-start* is the starting time in Epoch Time.
+   * *operating-mode* is the ending time in Epoch Time.
```
```suggestion
   * *localtime-end* is the ending time in Epoch Time.
```

kkasperczyk-no commented on samples/matter/lock/src/credentials/credentials_data_types.cpp:
```diff
@@ -176,4 +176,150 @@ CHIP_ERROR User::Deserialize(const void *buff, size_t buffSize)
 	return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR WeekDaySchedule::FillFromPlugin(EmberAfPluginDoorLockWeekDaySchedule &plugin)
```
1. Are schedules necessarily related to credentials? Now it looks a little bit like schedules would be a subset of credentials feature. Maybe we should work on some renaming e.g. to make this file describe generally all data types related to door lock features.
2. I think we should add `#ifdef CONFIG_LOCK_SCHEDULES` here to save some flash if schedules are not used.

ArekBalysNordic replied:
Ehh, we can but it requires changing in many places... We decided to have a credentials manager and now for example the CredentialsManager class is a template class that depends on CredentialsBits. So the whole module needs to be refactored then. I can move schedules to bolt_lock_manager, but then it will not be consistent with credentials_data_types, because they are still `credentials`. So what I can propose is to rename all modules and files from `credentials` to `lock_persistent`. @markaj-nordic  what do you think?

ArekBalysNordic replied:
2. done.

markaj-nordic replied:
To me, having the scheduled access feature implemented under the credentials module is quite confusing. My understanding is that, we have very similar implementation (because credentials and scheduled access operates in a similar manner and both needs persistence) of high level features that are quite unrelated. The easiest solution I see for now is to rename the CredentialsManager/Data to something more generic, like "AccessManager"/AccessData.

ArekBalysNordic replied:
f2f, decided to rename it to "Access"

kkasperczyk-no commented on samples/matter/lock/src/credentials/credentials_manager.h:
```diff
@@ -195,14 +213,37 @@ template <DoorLockData::CredentialsBits CRED_BIT_MASK> class CredentialsManager
 
 	void InitializeUsers();
-	void LoadFromPersistentStorage();
+	void InitializeSchedules();
```
Put this under ifdef below.

kkasperczyk-no commented on samples/matter/lock/src/credentials/credentials_manager_credentials.cpp:
```diff
@@ -0,0 +1,282 @@
+
+using namespace chip;
+using namespace DoorLockData;
+
```
Are the `credentials_manager_x.cpp` files added to the compilation anywhere?

kkasperczyk-no replied:
Btw. this is what I meant before about renaming the modules. Credentials manager credentials sounds weird.

ArekBalysNordic replied:
So as I wrote above, it can be transformed to "lock_persistent" for example.

kkasperczyk-no commented on samples/matter/lock/Kconfig:
```diff
@@ -32,6 +55,15 @@ config LOCK_ENABLE_DEBUG
 
 if LOCK_ENABLE_DEBUG
 
+config LOCK_PRINT_STORAGE_STATUS
```
Is it used anywhere? I can't find it.

ArekBalysNordic replied:
It is used in the `credentials_storage.cpp` please check one more time. Previously it was used but never defined in Kconfig. It is a fix for that.

kkasperczyk-no commented on samples/matter/lock/src/credentials/credentials_manager_schedules.cpp:
```diff
@@ -0,0 +1,372 @@
+ */
+
+#include "credentials_data_types.h"
+#ifdef CONFIG_LOCK_SCHEDULES
```
Wouldn't adding this file to compilation based on this Kconfig make more sense than putting the whole code under ifdef?

ArekBalysNordic replied:
Because we're using https://github.com/nrfconnect/sdk-nrf/blob/main/samples/matter/lock/CMakeLists.txt#L31 automatic source adding mechanism. @markaj-nordic You wrote this code before, do you think I should split them?

ArekBalysNordic replied:
I've split it.

markaj-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -536,6 +642,119 @@ After building the sample and programming it to your development kit, complete t
+   .. parsed-literal::
+      :class: highlight
+
+      ./chip-tool doorlock get-holiday-schedule *holiday-index* *destination-id* *endpoint-id*
```
Should we add some commends to actually test if the access works? Like opening the bolt lock in defined time slot?

markaj-nordic commented on samples/matter/lock/src/credentials/credentials_data_types.h:
```diff
@@ -118,6 +119,196 @@ struct Credential {
+		} mFields;
+		uint8_t mRaw[sizeof(mFields)];
+	};
+	static_assert(sizeof(Data) == 5);
```
Hmm without `packed` I'd expect the resulting size to be 8.

ArekBalysNordic replied:
Static assertions didn't point to failure.

markaj-nordic replied:
Ahh ok, probably because we have only 1 byte fields here.

markaj-nordic commented on samples/matter/lock/src/credentials/credentials_data_types.cpp:
```diff
@@ -176,4 +176,154 @@ CHIP_ERROR User::Deserialize(const void *buff, size_t buffSize)
+	}
+
+	if (buffSize > RequiredBufferSize()) {
+		return CHIP_ERROR_BUFFER_TOO_SMALL;
+	}
+
+	memcpy(mData.mRaw, buff, sizeof(mData.mRaw));
+
+	return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR YearDaySchedule::Deserialize(const void *buff, size_t buffSize)
+{
+	if (!buff) {
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	if (buffSize > RequiredBufferSize()) {
+		return CHIP_ERROR_BUFFER_TOO_SMALL;
+	}
+
+	memcpy(mData.mRaw, buff, sizeof(mData.mRaw));
+
+	return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR HolidaySchedule::Deserialize(const void *buff, size_t buffSize)
+{
+	if (!buff) {
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	if (buffSize > RequiredBufferSize()) {
+		return CHIP_ERROR_BUFFER_TOO_SMALL;
+	}
+
+	memcpy(mData.mRaw, buff, sizeof(mData.mRaw));
+
+	return CHIP_NO_ERROR;
+}
```
It's actually the same function. Maybe it could be templated somehow?

ArekBalysNordic replied:
It was templated previously, but we decided to have separate functions.

markaj-nordic commented on samples/matter/lock/src/credentials/credentials_data_types.h:
```diff
@@ -8,6 +8,7 @@
 
 #include <app/clusters/door-lock-server/door-lock-server.h>
 #include <lib/core/ClusterEnums.h>
+#include <type_traits>
```
Why is this one needed?

ArekBalysNordic replied:
It is not needed anymore, thanks for catching it.

markaj-nordic commented on samples/matter/lock/src/credentials/credentials_manager.h:
```diff
@@ -134,6 +134,19 @@ template <DoorLockData::CredentialsBits CRED_BIT_MASK> class CredentialsManager
 			   DlCredentialStatus credentialStatus, CredentialTypeEnum credentialType,
 			   const chip::ByteSpan &secret);
 
+	DlStatus GetWeekDaySchedule(uint8_t weekdayIndex, uint16_t userIndex,
```
Unfortunately, if we want to be consistent, all of these new methods should have doxygen commentary :)

ArekBalysNordic replied:
Done

markaj-nordic commented on samples/matter/lock/src/credentials/credentials_manager.h:
```diff
@@ -183,7 +196,12 @@ template <DoorLockData::CredentialsBits CRED_BIT_MASK> class CredentialsManager
 	const char *GetCredentialName(CredentialTypeEnum type);
 	void PrintUser(uint16_t userIndex);
-#endif
+#ifdef CONFIG_LOCK_SCHEDULES
```
Why does it apply only here?

ArekBalysNordic commented:
@kkasperczyk-no I've changed the naming from credentials to access, please review the recent changes.

markaj-nordic commented on samples/matter/lock/src/credentials/credentials_storage.cpp:
```diff
@@ -35,14 +35,18 @@ bool GetStorageFreeSpace(size_t &freeBytes)
 #define PSStore SecureStore
 #define PSLoad SecureLoad
+#else
+#error Please select only one of the possible credential backends: CONFIG_NCS_SAMPLE_MATTER_SECURE_STORAGE_BACKEND or CONFIG_NCS_SAMPLE_MATTER_SETTINGS_STORAGE_BACKEND
```
I'm not sure about this one. From the credential manager perspective this is true that we only need just one, but the application may use the other one for whatever reason, right?

ArekBalysNordic replied:
I've changed it to match your suggestion.

ArekBalysNordic commented:
> > @kkasperczyk-no I've changed the naming from credentials to access, please review the recent changes.
> 
> Did you check if we mention the module elsewhere in the documentation?

I found several places, however I hope I've replaced all of them.

wiba-nordic commented on samples/matter/lock/README.rst:
```diff
@@ -568,21 +568,56 @@ To enable the scheduled timed access feature, complete the following steps:
+
+#. In the **Time Synchronization** context window, go to the **Attributes** tab and set the proper bits for the ``FeatureMap`` attribute:
+
+   * For the ``TimeZone`` support set the ``0th`` bit of the feature map bit map.
```
```suggestion
   * For the ``TimeZone`` support, set the ``0th`` bit of the feature map bit map.
```
